File: src/main/java/itmo/is/project/mapper/user/UserMapper.java
package itmo.is.project.mapper.user;

import itmo.is.project.dto.security.AuthenticationRequest;
import itmo.is.project.dto.security.RegistrationRequest;
import itmo.is.project.dto.user.UserDto;
import itmo.is.project.mapper.EntityMapper;
import itmo.is.project.model.user.User;
import org.mapstruct.Mapper;

@Mapper(componentModel = "spring")
public interface UserMapper extends EntityMapper<UserDto, User> {
    User toEntity(RegistrationRequest registrationRequest);

    User toEntity(AuthenticationRequest authenticationRequest);
}

File: src/main/kotlin/itmo/isproject/mapper/user/UserMapper.kt
package itmo.isproject.mapper.user

import itmo.isproject.dto.security.AuthenticationRequest
import itmo.isproject.dto.security.RegistrationRequest
import itmo.isproject.dto.user.UserDto
import itmo.isproject.mapper.EntityMapper
import itmo.isproject.model.user.User
import org.mapstruct.*

@Mapper
interface UserMapper : EntityMapper<UserDto, User> {

    @Mappings(
        Mapping(source = "usernameInternal", target = "username"),
    )
    override fun toDto(entity: User): UserDto

    @InheritInverseConfiguration
    @Mapping(target = "usernameInternal", source = "username")
    override fun toEntity(dto: UserDto): User

    @Mapping(target = "passwordInternal", source = "password")
    @Mapping(target = "usernameInternal", source = "username")
    fun toEntity(registrationRequest: RegistrationRequest): User

    @Mapping(target = "usernameInternal", source = "username")
    @Mapping(target = "passwordInternal", source = "password")
    fun toEntity(authenticationRequest: AuthenticationRequest): User
}